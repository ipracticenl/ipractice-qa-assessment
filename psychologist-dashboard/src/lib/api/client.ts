import {
  Client,
  Appointment,
  Availability,
  CreateClientDto,
  CreateBookingDto,
  ClientDetailsDto,
  ClientSummaryDto,
} from "../../types";

const API_BASE_URL =
  import.meta.env.VITE_API_BASE_URL || "http://localhost:5000";

// Client API interfaces
export interface CreateClientData extends CreateClientDto {}

export interface CreateBookingData extends CreateBookingDto {}

export interface AvailableTimeSlotsResult {
  psychologistId: number;
  psychologistName: string;
  availableTimeSlots: Availability[];
}

// Create a new client
export async function createClient(data: CreateClientData): Promise<Client> {
  const response = await fetch(`${API_BASE_URL}/Client`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(data),
  });

  if (!response.ok) {
    throw new Error("Error creating client");
  }

  return response.json();
}

// Get client by ID
export async function getClient(id: number): Promise<ClientDetailsDto> {
  const response = await fetch(`${API_BASE_URL}/Client/${id}`);

  if (!response.ok) {
    throw new Error("Error fetching client");
  }

  return response.json();
}

// Get all clients
export async function getAllClients(): Promise<ClientSummaryDto[]> {
  const response = await fetch(`${API_BASE_URL}/Client`);

  if (!response.ok) {
    throw new Error("Error fetching clients");
  }

  return response.json();
}

// Get client by name
export async function getClientByName(name: string): Promise<Client> {
  const response = await fetch(
    `${API_BASE_URL}/Client/by-name/${encodeURIComponent(name)}`
  );

  if (!response.ok) {
    if (response.status === 404) {
      throw new Error("Client not found");
    }
    throw new Error("Error fetching client by name");
  }

  return response.json();
}

// Get available time slots for a client
export async function getAvailableTimeSlots(
  clientId: number
): Promise<AvailableTimeSlotsResult[]> {
  const response = await fetch(
    `${API_BASE_URL}/Client/${clientId}/available-timeslots`
  );

  if (!response.ok) {
    throw new Error("Error fetching available time slots");
  }

  return response.json();
}

// Book an appointment
export async function bookAppointment(
  clientId: number,
  data: CreateBookingData
): Promise<Appointment> {
  const response = await fetch(`${API_BASE_URL}/Client/${clientId}/bookings`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(data),
  });

  if (!response.ok) {
    throw new Error("Error booking appointment");
  }

  return response.json();
}

// Cancel an appointment
export async function cancelAppointment(
  clientId: number,
  appointmentId: string
): Promise<void> {
  const response = await fetch(
    `${API_BASE_URL}/Client/${clientId}/bookings/${appointmentId}`,
    {
      method: "DELETE",
    }
  );

  if (!response.ok) {
    throw new Error("Error canceling appointment");
  }
}
