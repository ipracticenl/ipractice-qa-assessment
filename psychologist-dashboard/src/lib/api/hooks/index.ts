// Export all client hooks
export * from "./use-client";

// Export all psychologist hooks
export * from "./use-psychologist";
