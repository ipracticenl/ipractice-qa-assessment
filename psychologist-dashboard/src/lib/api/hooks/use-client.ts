import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query";
import { toast } from "sonner";
import {
  createClient,
  getClient,
  getClientByName,
  getAllClients,
  getAvailableTimeSlots,
  bookAppointment,
  cancelAppointment,
  type CreateClientData,
  type CreateBookingData,
} from "../client";

// Hook to get a client by ID
export const useClient = (clientId?: number) => {
  return useQuery({
    queryKey: ["client", clientId],
    queryFn: () => getClient(clientId!),
    enabled: !!clientId,
  });
};

// Hook to get all clients
export const useAllClients = (enabled: boolean = true) => {
  return useQuery({
    queryKey: ["clients"],
    queryFn: () => getAllClients(),
    enabled,
  });
};

// Hook to get a client by name
export const useClientByName = (clientName?: string) => {
  return useQuery({
    queryKey: ["client", "by-name", clientName],
    queryFn: () => getClientByName(clientName!),
    enabled: !!clientName,
  });
};

// Hook to get available time slots for a client
export const useAvailableTimeSlots = (clientId?: number) => {
  return useQuery({
    queryKey: ["availableTimeSlots", clientId],
    queryFn: () => getAvailableTimeSlots(clientId!),
    enabled: !!clientId,
  });
};

// Hook to create a new client
export const useCreateClient = () => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (data: CreateClientData) => createClient(data),
    onSuccess: (newClient) => {
      // Invalidate and refetch client queries
      queryClient.invalidateQueries({ queryKey: ["client"] });
      // Optionally add the new client to the cache
      queryClient.setQueryData(["client", newClient.id], newClient);
      toast.success("Client created successfully!");
    },
    onError: () => {
      toast.error("Failed to create client. Please try again.");
    },
  });
};

// Hook to book an appointment
export const useBookAppointment = () => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: ({
      clientId,
      data,
    }: {
      clientId: number;
      data: CreateBookingData;
    }) => bookAppointment(clientId, data),
    onSuccess: (_, { clientId, data }) => {
      // Invalidate available time slots to reflect the booking
      queryClient.invalidateQueries({
        queryKey: ["availableTimeSlots", clientId],
      });
      queryClient.invalidateQueries({
        queryKey: ["client", clientId],
      });
      queryClient.invalidateQueries({
        queryKey: ["psychologists"],
      });
      queryClient.invalidateQueries({
        queryKey: ["psychologist", data.psychologistId],
      });
      // Invalidate any appointment-related queries
      queryClient.invalidateQueries({ queryKey: ["appointments"] });
      toast.success("Appointment booked successfully!");
    },
    onError: () => {
      toast.error("Failed to book appointment. Please try again.");
    },
  });
};

// Hook to cancel an appointment
export const useCancelAppointment = () => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: ({
      clientId,
      appointmentId,
    }: {
      clientId: number;
      appointmentId: string;
    }) => cancelAppointment(clientId, appointmentId),
    onSuccess: (_, { clientId }) => {
      // Invalidate available time slots to reflect the cancellation
      queryClient.invalidateQueries({
        queryKey: ["availableTimeSlots", clientId],
      });
      queryClient.invalidateQueries({
        queryKey: ["client", clientId],
      });

      // Invalidate any appointment-related queries
      queryClient.invalidateQueries({ queryKey: ["appointments"] });
      queryClient.invalidateQueries({ queryKey: ["psychologist"] });
      toast.success("Appointment cancelled successfully!");
    },
    onError: () => {
      toast.error("Failed to cancel appointment. Please try again.");
    },
  });
};
