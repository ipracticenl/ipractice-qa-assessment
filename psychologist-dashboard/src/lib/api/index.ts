export * from "./client";
export * from "./psychologist";

// Export all hooks
export * from "./hooks";
