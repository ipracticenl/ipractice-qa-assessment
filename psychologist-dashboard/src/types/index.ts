// Base types
export interface User {
  id: number;
  name: string;
  type: "psychologist" | "client";
}

// Time range base type
export interface TimeRange {
  id: string;
  from: Date;
  to: Date;
}

// Client DTOs
export interface CreateClientDto {
  name: string;
  initialPsychologistIds: number[];
}

export interface ClientSummaryDto {
  id: number;
  name: string;
}

export interface ClientDetailsDto {
  id: number;
  name: string;
  assignedPsychologists: number[];
  bookedAppointments: Appointment[];
}

// Psychologist DTOs
export interface CreatePsychologistDto {
  name: string;
  initialClients: number[];
}

export interface PsychologistSummaryDto {
  id: number;
  name: string;
}

export interface PsychologistDetailsDto {
  id: number;
  name: string;
  assignedClients: number[];
  availableTimeSlots: AvailableTimeSlot[];
  bookedAppointments: BookedAppointment[];
}

// Booking DTOs
export interface CreateBookingDto {
  psychologistId: number;
  availableTimeSlotId: string;
  comment?: string;
}

export interface ClientDto {
  clientId: number;
}

export interface TimeSlotDto {
  from: Date;
  to: Date;
}

// Model types
export interface AvailableTimeSlot extends TimeRange {
  // Inherits id, from, to from TimeRange
}

export interface Appointment extends TimeRange {
  psychologistId: number;
  comment?: string;
}

export interface BookedAppointment extends TimeRange {
  clientId: number;
}

// Legacy types for backward compatibility
export interface Psychologist extends User {
  type: "psychologist";
  clients: string[];
}

export interface Client extends User {
  type: "client";
  psychologist?: string;
}

export interface Availability {
  id: string;
  psychologistId: number;
  startTime: Date;
  endTime: Date;
  isBooked: boolean;
  clientId?: string;
  clientName?: string;
}
