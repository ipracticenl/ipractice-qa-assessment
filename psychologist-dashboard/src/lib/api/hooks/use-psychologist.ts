import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query";
import { toast } from "sonner";
import {
  getAllPsychologists,
  getPsychologistByName,
  createPsychologist,
  getPsychologist,
  assignClientToPsychologist,
  createAvailableTimeSlot,
  updateAvailableTimeSlot,
  deleteAvailableTimeSlot,
  type CreatePsychologistData,
  type AssignClientData,
  type TimeSlotData,
} from "../psychologist";
import { cancelAppointment } from "../client";

// Hook to get all psychologists
export const useAllPsychologists = (enabled: boolean = true) => {
  return useQuery({
    queryKey: ["psychologists"],
    queryFn: () => getAllPsychologists(),
    enabled,
  });
};

// Hook to get a psychologist by name
export const usePsychologistByName = (name?: string) => {
  return useQuery({
    queryKey: ["psychologist", "name", name],
    queryFn: () => getPsychologistByName(name!),
    enabled: !!name,
  });
};

// Hook to get a psychologist by ID
export const usePsychologist = (psychologistId?: number) => {
  return useQuery({
    queryKey: ["psychologist", psychologistId],
    queryFn: () => getPsychologist(psychologistId!),
    enabled: !!psychologistId,
  });
};

// Hook to create a new psychologist
export const useCreatePsychologist = () => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (data: CreatePsychologistData) => createPsychologist(data),
    onSuccess: (newPsychologist) => {
      // Invalidate and refetch psychologist queries
      queryClient.invalidateQueries({ queryKey: ["psychologist"] });
      // Optionally add the new psychologist to the cache
      queryClient.setQueryData(
        ["psychologist", newPsychologist.id],
        newPsychologist
      );
      toast.success("Psychologist created successfully!");
    },
    onError: () => {
      toast.error("Failed to create psychologist. Please try again.");
    },
  });
};

// Hook to assign a client to a psychologist
export const useAssignClientToPsychologist = () => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: ({
      psychologistId,
      data,
    }: {
      psychologistId: number;
      data: AssignClientData;
    }) => assignClientToPsychologist(psychologistId, data),
    onSuccess: (updatedPsychologist, { psychologistId }) => {
      // Update the psychologist in the cache
      queryClient.setQueryData(
        ["psychologist", psychologistId],
        updatedPsychologist
      );
      // Invalidate client queries as they might have changed
      queryClient.invalidateQueries({ queryKey: ["client"] });
      toast.success("Client assigned successfully!");
    },
    onError: () => {
      toast.error("Failed to assign client. Please try again.");
    },
  });
};

// Hook to create an available time slot for a psychologist
export const useCreateAvailableTimeSlot = () => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: ({
      psychologistId,
      data,
    }: {
      psychologistId: number;
      data: TimeSlotData;
    }) => createAvailableTimeSlot(psychologistId, data),
    onSuccess: (updatedPsychologist, { psychologistId }) => {
      // Update the psychologist in the cache
      queryClient.setQueryData(
        ["psychologist", psychologistId],
        updatedPsychologist
      );
      // Invalidate available time slots queries
      queryClient.invalidateQueries({ queryKey: ["availableTimeSlots"] });
      toast.success("Time slot created successfully!");
    },
    onError: () => {
      toast.error("Failed to create time slot. Please try again.");
    },
  });
};

// Hook to update an available time slot for a psychologist
export const useUpdateAvailableTimeSlot = () => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: ({
      psychologistId,
      timeSlotId,
      data,
    }: {
      psychologistId: number;
      timeSlotId: string;
      data: TimeSlotData;
    }) => updateAvailableTimeSlot(psychologistId, timeSlotId, data),
    onSuccess: (updatedPsychologist, { psychologistId }) => {
      // Update the psychologist in the cache
      queryClient.setQueryData(
        ["psychologist", psychologistId],
        updatedPsychologist
      );
      // Invalidate available time slots queries
      queryClient.invalidateQueries({ queryKey: ["availableTimeSlots"] });
      toast.success("Time slot updated successfully!");
    },
    onError: () => {
      toast.error("Failed to update time slot. Please try again.");
    },
  });
};

// Hook to cancel an session
export const useCancelSession = () => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: ({
      clientId,
      appointmentId,
    }: {
      clientId: number;
      appointmentId: string;
      psychologistId: number;
    }) => cancelAppointment(clientId, appointmentId),
    onSuccess: () => {
      const invalidateQueries = async () => {
        // Invalidate available time slots to reflect the cancellation
        queryClient.invalidateQueries({
          queryKey: ["psychologist"],
        });

        // Invalidate any appointment-related queries
        queryClient.invalidateQueries({ queryKey: ["appointments"] });

        toast.success("Session cancelled successfully!");
      };

      if (Math.random() < 0.5) {
        setTimeout(() => {
          invalidateQueries();
        }, 5000);
      } else {
        invalidateQueries();
      }
    },
    onError: () => {
      toast.error("Failed to cancel session. Please try again.");
    },
  });
};

// Hook to delete an available time slot for a psychologist
export const useDeleteAvailableTimeSlot = () => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: ({
      psychologistId,
      timeSlotId,
    }: {
      psychologistId: number;
      timeSlotId: string;
    }) => deleteAvailableTimeSlot(psychologistId, timeSlotId),
    onSuccess: (updatedPsychologist, { psychologistId }) => {
      // Update the psychologist in the cache
      queryClient.setQueryData(
        ["psychologist", psychologistId],
        updatedPsychologist
      );
      // Invalidate available time slots queries
      queryClient.invalidateQueries({ queryKey: ["availableTimeSlots"] });
      toast.success("Time slot deleted successfully!");
    },
    onError: () => {
      toast.error("Failed to delete time slot. Please try again.");
    },
  });
};
