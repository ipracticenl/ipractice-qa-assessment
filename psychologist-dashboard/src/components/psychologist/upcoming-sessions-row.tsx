import { UserMinus } from "lucide-react";
import { format } from "date-fns";
import type { FC } from "react";
import { Button } from "../ui/button";
import { TableRow, TableCell } from "../ui/table";
import { BookedAppointment } from "@/types";
import { useClient } from "@/lib/api/hooks";

interface UpcomingSessionsRowProps {
  session: BookedAppointment;
  onCancelSession: (sessionId: string, clientId: number) => void;
}

export const UpcomingSessionsRow: FC<UpcomingSessionsRowProps> = ({
  session,
  onCancelSession,
}) => {
  const { data: client } = useClient(session.clientId);
  return (
    <TableRow data-testid={`session-row-${session.id}`}>
      <TableCell className="font-medium">
        <div className="flex flex-col">
          <span>{format(session.from, "MM/dd/yyyy")}</span>
          <span className="text-sm text-gray-600">
            {format(session.from, "HH:mm")} - {format(session.to, "HH:mm")}
          </span>
        </div>
      </TableCell>
      <TableCell>{client?.name}</TableCell>
      <TableCell>
        <Button
          size="sm"
          variant="destructive"
          onClick={() => onCancelSession(session.id, session.clientId)}
          className="min-h-[36px] touch-manipulation"
          data-testid={`remove-session-${session.id}`}
          aria-label={`Cancel session with ${client?.name} on ${format(
            session.from,
            "MM/dd/yyyy HH:mm"
          )}`}
        >
          <UserMinus className="w-3 h-3 mr-1" />
          Cancel
        </Button>
      </TableCell>
    </TableRow>
  );
};
