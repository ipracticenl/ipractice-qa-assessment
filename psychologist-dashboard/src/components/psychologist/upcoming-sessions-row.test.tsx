import { render, screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi, beforeEach } from "vitest";
import { UpcomingSessionsRow } from "./upcoming-sessions-row";
import { BookedAppointment } from "@/types";

// Mock the useClient hook
vi.mock("@/lib/api/hooks", () => ({
  useClient: vi.fn(),
}));

// Import the mocked hook
import { useClient } from "@/lib/api/hooks";

// Mock data
const mockSession: BookedAppointment = {
  id: "session-1",
  from: new Date("2024-01-15T10:00:00"),
  to: new Date("2024-01-15T11:00:00"),
  clientId: 1,
};

const mockClient = {
  id: 1,
  name: "John Doe",
};

describe("UpcomingSessionsRow", () => {
  const mockOnCancelSession = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe("Snapshot Tests", () => {
    it("should match snapshot when client data is loaded", () => {
      // Given: Client data is available
      vi.mocked(useClient).mockReturnValue({
        data: mockClient,
        isLoading: false,
        error: null,
      } as any);

      // When: The component is rendered
      const { container } = render(
        <UpcomingSessionsRow
          session={mockSession}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The component should match the snapshot
      expect(container).toMatchSnapshot();
    });

    it("should match snapshot when client data is loading", () => {
      // Given: Client data is loading
      vi.mocked(useClient).mockReturnValue({
        data: null,
        isLoading: true,
        error: null,
      } as any);

      // When: The component is rendered
      const { container } = render(
        <UpcomingSessionsRow
          session={mockSession}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The component should match the snapshot
      expect(container).toMatchSnapshot();
    });
  });

  describe("Session Display", () => {
    it("should display session date and time correctly", () => {
      // Given: Client data is available
      vi.mocked(useClient).mockReturnValue({
        data: mockClient,
        isLoading: false,
        error: null,
      } as any);

      // When: The component is rendered
      render(
        <UpcomingSessionsRow
          session={mockSession}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The date and time should be formatted correctly
      expect(screen.getByText("01/15/2024")).toBeInTheDocument();
      expect(screen.getByText("10:00 - 11:00")).toBeInTheDocument();
    });

    it("should display client name when available", () => {
      // Given: Client data is available
      vi.mocked(useClient).mockReturnValue({
        data: mockClient,
        isLoading: false,
        error: null,
      } as any);

      // When: The component is rendered
      render(
        <UpcomingSessionsRow
          session={mockSession}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The client name should be displayed
      expect(screen.getByText("John Doe")).toBeInTheDocument();
    });

    it("should handle missing client data gracefully", () => {
      // Given: Client data is not available
      vi.mocked(useClient).mockReturnValue({
        data: null,
        isLoading: false,
        error: null,
      } as any);

      // When: The component is rendered
      render(
        <UpcomingSessionsRow
          session={mockSession}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The component should render without client name
      expect(screen.queryByText("John Doe")).not.toBeInTheDocument();
    });
  });

  describe("Cancel Functionality", () => {
    it("should call onCancelSession when remove button is clicked", () => {
      // Given: Client data is available and cancel function is mocked
      vi.mocked(useClient).mockReturnValue({
        data: mockClient,
        isLoading: false,
        error: null,
      } as any);
      const mockCancelFunction = vi.fn();

      // When: The component is rendered and remove button is clicked
      render(
        <UpcomingSessionsRow
          session={mockSession}
          onCancelSession={mockCancelFunction}
        />
      );

      const removeButton = screen.getByTestId("remove-session-session-1");
      fireEvent.click(removeButton);

      // Then: The cancel function should be called with correct parameters
      expect(mockCancelFunction).toHaveBeenCalledWith("session-1", 1);
    });

    it("should have correct aria-label for remove button", () => {
      // Given: Client data is available
      vi.mocked(useClient).mockReturnValue({
        data: mockClient,
        isLoading: false,
        error: null,
      } as any);

      // When: The component is rendered
      render(
        <UpcomingSessionsRow
          session={mockSession}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The remove button should have the correct aria-label
      const removeButton = screen.getByTestId("remove-session-session-1");
      expect(removeButton).toHaveAttribute("aria-label");
      expect(removeButton.getAttribute("aria-label")).toContain(
        "Cancel session with John Doe on 01/15/2024 10:00"
      );
    });

    it("should handle aria-label when client name is not available", () => {
      // Given: Client data is not available
      vi.mocked(useClient).mockReturnValue({
        data: null,
        isLoading: false,
        error: null,
      } as any);

      // When: The component is rendered
      render(
        <UpcomingSessionsRow
          session={mockSession}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The remove button should still have an aria-label
      const removeButton = screen.getByTestId("remove-session-session-1");
      expect(removeButton).toHaveAttribute("aria-label");
      expect(removeButton.getAttribute("aria-label")).toContain(
        "Cancel session with undefined on 01/15/2024 10:00"
      );
    });
  });

  describe("Component Structure", () => {
    it("should render the row with correct test ID", () => {
      // Given: Client data is available
      vi.mocked(useClient).mockReturnValue({
        data: mockClient,
        isLoading: false,
        error: null,
      } as any);

      // When: The component is rendered
      render(
        <UpcomingSessionsRow
          session={mockSession}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The row should be rendered with correct test ID
      expect(screen.getByTestId("session-row-session-1")).toBeInTheDocument();
    });

    it("should render remove button with correct styling", () => {
      // Given: Client data is available
      vi.mocked(useClient).mockReturnValue({
        data: mockClient,
        isLoading: false,
        error: null,
      } as any);

      // When: The component is rendered
      render(
        <UpcomingSessionsRow
          session={mockSession}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The remove button should have correct styling classes
      const removeButton = screen.getByTestId("remove-session-session-1");
      expect(removeButton).toHaveClass("min-h-[36px]", "touch-manipulation");
    });

    it('should display "Remove" text on button', () => {
      // Given: Client data is available
      vi.mocked(useClient).mockReturnValue({
        data: mockClient,
        isLoading: false,
        error: null,
      } as any);

      // When: The component is rendered
      render(
        <UpcomingSessionsRow
          session={mockSession}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The button should display "Remove" text
      expect(screen.getByText("Cancel")).toBeInTheDocument();
    });
  });

  describe("Hook Integration", () => {
    it("should call useClient with correct client ID", () => {
      // Given: The component is about to be rendered
      vi.mocked(useClient).mockReturnValue({
        data: mockClient,
        isLoading: false,
        error: null,
      } as any);

      // When: The component is rendered
      render(
        <UpcomingSessionsRow
          session={mockSession}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: useClient should be called with the correct client ID
      expect(useClient).toHaveBeenCalledWith(1);
    });

    it("should handle loading state", () => {
      // Given: Client data is loading
      vi.mocked(useClient).mockReturnValue({
        data: null,
        isLoading: true,
        error: null,
      } as any);

      // When: The component is rendered
      render(
        <UpcomingSessionsRow
          session={mockSession}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The component should render without client name
      expect(screen.queryByText("John Doe")).not.toBeInTheDocument();
    });

    it("should handle error state", () => {
      // Given: Client data has an error
      vi.mocked(useClient).mockReturnValue({
        data: null,
        isLoading: false,
        error: new Error("Failed to load client"),
      } as any);

      // When: The component is rendered
      render(
        <UpcomingSessionsRow
          session={mockSession}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The component should render without client name
      expect(screen.queryByText("John Doe")).not.toBeInTheDocument();
    });
  });

  describe("Edge Cases", () => {
    it("should handle different date formats correctly", () => {
      // Given: A session with different date
      const differentSession: BookedAppointment = {
        id: "session-2",
        from: new Date("2024-12-25T15:30:00"),
        to: new Date("2024-12-25T16:30:00"),
        clientId: 2,
      };

      vi.mocked(useClient).mockReturnValue({
        data: { id: 2, name: "Jane Smith" },
        isLoading: false,
        error: null,
      } as any);

      // When: The component is rendered
      render(
        <UpcomingSessionsRow
          session={differentSession}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The date should be formatted correctly
      expect(screen.getByText("12/25/2024")).toBeInTheDocument();
      expect(screen.getByText("15:30 - 16:30")).toBeInTheDocument();
    });

    it("should handle empty client name", () => {
      // Given: Client data with empty name
      vi.mocked(useClient).mockReturnValue({
        data: { id: 1, name: "" },
        isLoading: false,
        error: null,
      } as any);

      // When: The component is rendered
      render(
        <UpcomingSessionsRow
          session={mockSession}
          onCancelSession={mockOnCancelSession}
        />
      );

      // Then: The component should handle empty name gracefully
      const removeButton = screen.getByTestId("remove-session-session-1");
      expect(removeButton.getAttribute("aria-label")).toContain(
        "Cancel session with  on 01/15/2024 10:00"
      );
    });
  });
});
