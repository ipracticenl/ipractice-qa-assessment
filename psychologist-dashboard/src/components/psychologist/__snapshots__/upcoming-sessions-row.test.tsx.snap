// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UpcomingSessionsRow > Snapshot Tests > should match snapshot when client data is loaded 1`] = `
<div>
  <tr
    class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
    data-testid="session-row-session-1"
  >
    <td
      class="p-4 align-middle [&:has([role=checkbox])]:pr-0 font-medium"
    >
      <div
        class="flex flex-col"
      >
        <span>
          01/15/2024
        </span>
        <span
          class="text-sm text-gray-600"
        >
          10:00
           - 
          11:00
        </span>
      </div>
    </td>
    <td
      class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
    >
      John Doe
    </td>
    <td
      class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
    >
      <button
        aria-label="Cancel session with John Doe on 01/15/2024 10:00"
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-destructive text-destructive-foreground hover:bg-destructive/90 h-9 rounded-md px-3 min-h-[36px] touch-manipulation"
        data-testid="remove-session-session-1"
      >
        <svg
          class="lucide lucide-user-minus w-3 h-3 mr-1"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 21v-2a4 4 0 0 0-4-4H6a4 4 0 0 0-4 4v2"
          />
          <circle
            cx="9"
            cy="7"
            r="4"
          />
          <line
            x1="22"
            x2="16"
            y1="11"
            y2="11"
          />
        </svg>
        Cancel
      </button>
    </td>
  </tr>
</div>
`;

exports[`UpcomingSessionsRow > Snapshot Tests > should match snapshot when client data is loading 1`] = `
<div>
  <tr
    class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
    data-testid="session-row-session-1"
  >
    <td
      class="p-4 align-middle [&:has([role=checkbox])]:pr-0 font-medium"
    >
      <div
        class="flex flex-col"
      >
        <span>
          01/15/2024
        </span>
        <span
          class="text-sm text-gray-600"
        >
          10:00
           - 
          11:00
        </span>
      </div>
    </td>
    <td
      class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
    />
    <td
      class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
    >
      <button
        aria-label="Cancel session with undefined on 01/15/2024 10:00"
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-destructive text-destructive-foreground hover:bg-destructive/90 h-9 rounded-md px-3 min-h-[36px] touch-manipulation"
        data-testid="remove-session-session-1"
      >
        <svg
          class="lucide lucide-user-minus w-3 h-3 mr-1"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 21v-2a4 4 0 0 0-4-4H6a4 4 0 0 0-4 4v2"
          />
          <circle
            cx="9"
            cy="7"
            r="4"
          />
          <line
            x1="22"
            x2="16"
            y1="11"
            y2="11"
          />
        </svg>
        Cancel
      </button>
    </td>
  </tr>
</div>
`;
